### PR TITLE
Flush error_logger before capturing

### PR DESCRIPTION
**Pull request by fishcakez**
/cc @myronmarston

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 1]
```diff
@@ -37,7 +37,14 @@ defmodule ExUnit.CaptureLog do
   Returns the binary which is the captured output.
 
   This function mutes the `:console` backend
-  and captures any log messages sent to Logger.
+  and captures any log messages sent to Logger from the calling
+  processes. It is possible to ensure explicit log messages from other
+  processes are captured by waiting for their exit or monitor
+  signal. This does not guarantee to capture VM generated log messages
+  when a process spawned using a `spawn` BIF does not catch an exception.
```
Comment (myronmarston) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: I've definitely heard the term BIF used in erlang docs, but don't think I've encountered it in the Elixir docs so I've assumed it wasn't really a term used by Elixir.  Should this be `Kernel.spawn` or similar instead?

**Comment by myronmarston** [comment]
:+1:  Thanks, @fishcakez!

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 2]
```diff
@@ -37,7 +37,14 @@ defmodule ExUnit.CaptureLog do
...
   This function mutes the `:console` backend
-  and captures any log messages sent to Logger.
+  and captures any log messages sent to Logger from the calling
+  processes. It is possible to ensure explicit log messages from other
+  processes are captured by waiting for their exit or monitor
+  signal. This does not guarantee to capture VM generated log messages
+  when a process is spawned using a low level `Kernel` spawn function (e.g.
+  `Kernel.spawn/1`) does not catch an exception. A `Task`, or other OTP
+  process, will send explicit logs that are sent before its exit or monitor
+  signal and will not cause VM generated log messages.
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: @myronmarston is this description clearer?

Comment (myronmarston) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: It's better, but I think the "does not catch an exception" part is a bit confusing.  This spawned process does not catch an exception, but I don't think it's what you mean:

``` elixir
spawn(fn -> :ok end)
```

Maybe "raises an unhandled exception" or just "crashes" or something similar would be clearer?

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 3]
```diff
@@ -37,7 +37,15 @@ defmodule ExUnit.CaptureLog do
...
   This function mutes the `:console` backend
-  and captures any log messages sent to Logger.
+  and captures any log messages sent to Logger from the calling
+  processes. It is possible to ensure explicit log messages from other
+  processes are captured by waiting for their exit or monitor
+  signal. This does not guarantee to capture VM generated log messages
+  when a process is spawned using a low level `Kernel` spawn function (e.g.
+  `Kernel.spawn/1`) raises an exception that isn't rescued or a thow that
+  isn't caught. A `Task`, orther OTP process, will send explicit logs that
+  are sent before its exit or monitor signal and will not cause VM generated
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: @myronmarston how about this rewording?

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 4]
```diff
@@ -37,7 +37,15 @@ defmodule ExUnit.CaptureLog do
   Returns the binary which is the captured output.
 
   This function mutes the `:console` backend
-  and captures any log messages sent to Logger.
+  and captures any log messages sent to Logger from the calling
+  processes. It is possible to ensure explicit log messages from other
+  processes are captured by waiting for their exit or monitor
+  signal. This does not guarantee to capture VM generated log messages
+  when a process is spawned using a low level `Kernel` spawn function (e.g.
+  `Kernel.spawn/1`) raises an exception that isn't rescued or a thow that
```
Comment (myronmarston) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: s/thow/throw/

Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: Fixed

**Comment by fishcakez** [comment]
Please don't squash/merge before #5262 and please do backport to `v1.3` once merged.

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 5]
```diff
@@ -37,7 +37,15 @@ defmodule ExUnit.CaptureLog do
   Returns the binary which is the captured output.
 
   This function mutes the `:console` backend
-  and captures any log messages sent to Logger.
+  and captures any log messages sent to Logger from the calling
+  processes. It is possible to ensure explicit log messages from other
+  processes are captured by waiting for their exit or monitor
+  signal. This does not guarantee to capture VM generated log messages
+  when a process is spawned using a low level `Kernel` spawn function (e.g.
```
Comment (myronmarston) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: Grammatically, "process is spawned...raises" is awkward.  "process that is spawned...raises" reads better.

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 6]
```diff
@@ -37,7 +37,15 @@ defmodule ExUnit.CaptureLog do
...
+  `Kernel.spawn/1`) raises an exception that isn't rescued or a throw that
+  isn't caught. A `Task`, orther OTP process, will send explicit logs that
```
Comment (myronmarston) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: s/orther/or other/

File: lib/ex_unit/lib/ex_unit/capture_log.ex (hunk 3 above)
Comment (myronmarston) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: Much clearer.  I found a couple more grammar/spelling issues.  I'm happy with it once those are addressed :).

**Comment by fishcakez** [comment]
English is my native language :heart:

File: lib/ex_unit/lib/ex_unit/capture_log.ex (hunk 5 above)
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: fixed

File: lib/ex_unit/lib/ex_unit/capture_log.ex (hunk 6 above)
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_log.ex]: fixed

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**